plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/SpecialKotlinMethodEmbedding.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.conversion
+
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.formver.domains.UnitDomain
+import org.jetbrains.kotlin.formver.embeddings.*
+import org.jetbrains.kotlin.formver.viper.MangledName
+import org.jetbrains.kotlin.formver.viper.ast.Exp
+import org.jetbrains.kotlin.formver.viper.ast.Stmt
+import org.jetbrains.kotlin.name.CallableId
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+abstract class SpecialKotlinMethodEmbedding : MethodEmbedding {
+    override val preconditions: List<Exp> = listOf()
+    override val postconditions: List<Exp> = listOf()
+    override val viperMethod = null
+
+    abstract val packageSegments: List<String>
+    abstract val className: String?
+    abstract val methodName: String
+    override val name: MangledName
+        get() = CallableId(FqName.fromSegments(packageSegments), className?.let { FqName(it) }, Name.identifier(methodName)).embedName()
+}
+
+object KotlinContractMethodEmbedding : SpecialKotlinMethodEmbedding() {
+    override val packageSegments: List<String> = listOf("kotlin", "contracts")
+    override val className: String? = null
+    override val methodName: String = "contract"
+
+    override fun insertCall(argsFir: List<FirExpression>, ctx: StmtConversionContext<ResultTrackingContext>): Exp = UnitDomain.element
+
+    override val receiverType: TypeEmbedding? = null
+    override val paramTypes: List<TypeEmbedding> = listOf(FunctionTypeEmbedding)
+    override val returnType: TypeEmbedding = UnitTypeEmbedding
+}
+
+abstract class ExpressionShapedMethodEmbedding : SpecialKotlinMethodEmbedding() {
+    abstract fun callExpression(args: List<Exp>): Exp
+    open val callExpressionType: TypeEmbedding
+        get() = returnType
+
+    override fun insertCall(argsFir: List<FirExpression>, ctx: StmtConversionContext<ResultTrackingContext>): Exp =
+        callWithConversions(argsFir, ctx) { Pair(callExpression(it), callExpressionType) }
```

Comment:
Note that at least in our cases, the conversion is technically unnecessary. Nothing can inherit from `Int` or `Boolean` and the return values are also just converted to the same type.

Reply:
That's true for now, but consider for example `x.xor(y)` where `y: Nothing`.  This should again be fixed by the typed embedding, though.